third-party/CMakeLists.txt

```diff
@@ -50,6 +51,62 @@ function(osqueryThirdpartyMain)
   add_subdirectory("zstd")
 endfunction()
 
+function(osqueryThirdpartyBuildMain)
+  # For our custom ExternalProject wrapper we define argument groups.
+  add_custom_target(thirdparty_prebuilt)
+endfunction()
+
+function(definePrebuildProject name)
+  include(ExternalProject)
+  set(_ep_keywords_definePrebuildProject "^(URL|SHA256|DEPENDS|CONFIGURE|BUILD|INSTALL|BUILD_IN_SOURCE)$")
+
+  set(identifier "thirdparty_build_${name}")
+
+  # Create a target to hold the arguments from the function call.
+  set(props "${identifier}_props")
+  add_custom_target(${props})
+  _ep_parse_arguments(definePrebuildProject ${props} _CEP_ "${ARGN}")
+
+  # Access arguments as propertoes.
+  get_property(depends_list TARGET ${props} PROPERTY _CEP_DEPENDS)
+  foreach(depends ${depends_list})
+    list(APPEND depends_targets "thirdparty_build_${depends}")
+  endforeach()
+
+  get_property(hash TARGET ${props} PROPERTY _CEP_SHA256)
+  get_property(configure TARGET ${props} PROPERTY _CEP_CONFIGURE)
+  get_property(build TARGET ${props} PROPERTY _CEP_BUILD)
+  get_property(install TARGET ${props} PROPERTY _CEP_INSTALL)
+  get_property(build_in_source TARGET ${props} PROPERTY _CEP_BUILD_IN_SOURCE)
+
+  if(DEFINED PLATFORM_LINUX)
+    set(platform_name "linux")
+  elseif(DEFINED PLATFORM_MACOS)
+    set(platform_name "macos")
+  elseif(DEFINED PLATFORM_WINDOWS)
+    set(platform_name "windows")
+  else()
+    message(FATAL_ERROR "Unrecognized system")
+    return()
+  endif()
+
+  set(install_prefix "${THIRD_PARTY_PREBUILT_PATH}/${platform_name}-x86_64")
+
+  # 'Forward' them to ExternalProject
+  ExternalProject_Add(${identifier}
+    URL ${url}
+    INSTALL_DIR "${install_prefix}"
+    #HASH ${hash}
+    DEPENDS ${depends_targets}
+    CONFIGURE_COMMAND ${configure}
```

Comment:
We require a CONFIGURE/BUILD/INSTALL, instead of holding all commands in a single step because some projects will need specific logic for a platform. This means we'll need to add a `ExternalProject_Add_Step` that comes before BUILD or before INSTALL. 

We can ignore the "patch" step concept and patch within configure. I actually do not like the patch step within external project because it, and the other steps, are not idempotent. If a project fails to build in any way I would hope the retry starts fresh, but instead it started where it left off.

Reply:
I'm not sure I understand what's the problem with the PATCH step.
Can you make a specific example?

I mean I see several problems and things that could happen in every step, but I'm not sure what you're trying to achieve by moving the patch in the configure step.

Reply:
The problem I ran into was idempotent steps, say that `PATCH` does several actions. If one of those fail and you go to the `CMakeLists.txt` and fix it, the next invocation will re-run all of `PATCH` from the 'dirty' location it was running before. This would try to patch already-patched files.

I saw now benefit of having separate CONFIGURE and PATCH steps without some extra power of reverting the state on-failure.

But that said, I introduced a generic "apply patchfiles" step that abstracts the `PATCH_COMMAND` nicely. I'll post this shortly.